### PR TITLE
transport JavaScript standard builtin objects

### PR DESCRIPTION
**Pull request by helloshuangzi**

### CODE REVIEW

File: src/transport/deserializer.h [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+#pragma once
+
+#include "serialized-data.h"
+
+#include <napa/module/binding/wraps.h>
+
+namespace napa {
+namespace transport {
+
+    using namespace v8;
```
Comment (daiyip) [inline src/transport/deserializer.h]: Please don't put `using namespace` in header file.

File: src/transport/externalized-contents.h [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+// Licensed under the MIT license.
+
+#pragma once
+
+#include <v8.h>
+
+namespace napa {
+namespace transport {
+
+    using namespace v8;
```
Comment (daiyip) [inline src/transport/externalized-contents.h]: the same

File: src/transport/externalized-contents.h [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+    class ExternalizedContents {
```
Comment (daiyip) [inline src/transport/externalized-contents.h]: Add summary for this class.

File: src/transport/deserializer.h [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+    class Deserializer : public ValueDeserializer::Delegate {
```
Comment (daiyip) [inline src/transport/deserializer.h]: Add summary.

File: src/transport/deserializer.h [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        Isolate* _isolate;
+        ValueDeserializer _deserializer;
+        std::shared_ptr<SerializedData> _data;
+
+        Deserializer(const Deserializer&) = delete;
+        Deserializer& operator=(const Deserializer&) = delete;
+    };
+
+
+    Deserializer::Deserializer(Isolate* isolate, std::shared_ptr<SerializedData> data) :
```
Comment (daiyip) [inline src/transport/deserializer.h]: Still cannot put into CPP file?

File: src/transport/serialized-data.h [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+#include "externalized-contents.h"
+
+namespace napa {
+namespace transport {
+
+    using namespace v8;
+
+    typedef std::pair<SharedArrayBuffer::Contents, std::shared_ptr<ExternalizedContents>> ExternalizedSharedArrayBufferContents;
+
+    class SerializedData {
```
Comment (daiyip) [inline src/transport/serialized-data.h]: Add summary.

File: src/transport/serializer.h [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+#pragma once
+
+#include "serialized-data.h"
+
+#include <napa/module/binding/wraps.h>
+
+namespace napa {
+namespace transport {
+
+    using namespace v8;
```
Comment (daiyip) [inline src/transport/serializer.h]: the same.

File: src/transport/serialized-data.h [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+// Licensed under the MIT license.
+
+#pragma once
+
+#include "externalized-contents.h"
+
+namespace napa {
+namespace transport {
+
+    using namespace v8;
```
Comment (daiyip) [inline src/transport/serialized-data.h]: the same.

File: src/transport/transport-utils.h [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+
+#pragma once
+
+#include "deserializer.h"
+#include "serializer.h"
+
+namespace napa {
+namespace transport {
+
+    using namespace v8;
```
Comment (daiyip) [inline src/transport/transport-utils.h]: the same.

File: test/transport-builtin-test.js [hunk 10]
```diff
@@ -0,0 +1,246 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+const assert = require('assert');
+const napa = require('../lib/index');
+let zoneKey = 'transport-test-zone';
```
Comment (daiyip) [inline test/transport-builtin-test.js]: zoneId?

File: test/transport-test.ts [hunk 11]
```diff
@@ -5,6 +5,7 @@ import * as napa from "../lib/index";
 import * as assert from 'assert';
 import * as path from 'path';
 import * as t from './napa-zone/test';
+var transportBuiltin = require('./transport-builtin-test');
```
Comment (daiyip) [inline test/transport-test.ts]: import * as xxx from '...';

**Comment by helloshuangzi** [comment]
Both node and v8 are compiled with -fno-rtti (Run Time Type Infomation), but napajs isn't.
With -frtti (Default), any c++ class with virtual members will be compiled with the 2 symbols, vtable and typeinfo; while with -fno-rtti, the symbol of typeinfo will be optimized out.

Before updating the config for napa, napa-binding.node and napa.dll will try to retrieve typeinfo of v8::(De)Serializer::Delegate from Node, but that of Node has been optimized out. The dynamic-link error "undefined symbols" occurs here.

If we set -fno-rtti to compile napa::transport::(De)Serializer which are deprived from  v8::(De)Serializer::Delegate, both napa and node/v8 share the same level of optimization with the same way of symbol reference, the dynamic-link success.

At the same time, -fno-rtti means some type meta-info is optimized out, so you can't do some type manipulation, like "dynamic_cast". "inc/napa/memory/allocator-debugger.h" and "third-party/args/args.hxx" have dynamic_cast logic, that's why [the build](https://travis-ci.org/Microsoft/napajs/jobs/317278044) failed .

Propose::
1) a) To minimize the impact, for this change, I would like to make src/v8-extension a static lib to accommodate v8 extensions, including ArrayBufferAllocator/Serializer/Deserializer, and give -fno-rtti to this lib, while keep the remaining part of napa as it is now (-frtti). 
b) In the future, we would be aware that any v8 extensions required to be compiled with -fno-rtti need to go to this static lib. This static lib serves as a symbol coordinator between napa and v8. 
c) We may also need this kind of lib for node-extension in the future.
d) Keep napa::TransportUtils in src/transport.

2) For consistency with node/v8 (maybe aggressive but has benefits), we may consider to restrict those type manipulation in napa logic by default. It mean set -fno-rtti for napa by default. If some logic need it, just make that part specially have -frtti.

**Comment by daiyip** [comment]
Agree with option 1.

**Comment by helloshuangzi** [comment]
1. Resolved comments,
2. Created #149 as P2, and #150 as P1, will work on them as separate changes.
3. As to API documentation, 
- For JS builtin types,  because this solution is by design transparent for users to transport them, no new API is introduced.  
- For user defined type, (de)serializeValue of lib/transport/builtin-object-transporter.ts could be used to implement Transportable object. 
Will add api docs for the above parts together with examples in a separate change.

**Comment by daiyip** [comment]
Great to see this change eventually work on Linux/Mac.

I insist that we should unify .cpp and .cc as one convention since there is no reason to have them separate. I understand that v8/node uses .cc convention, but v8_extensions are part of napa code, I think it's better to follow napa's convention.

**Pull request merged**